qa/rpc-tests/wallet_protectcoinbase.py

```diff
@@ -0,0 +1,128 @@
+
+        assert_equal(self.nodes[0].getbalance(), 40)
+        assert_equal(self.nodes[1].getbalance(), 10)
+        assert_equal(self.nodes[2].getbalance(), 0)
+
+        # Send will fail because we are enforcing the consensus rule that
+        # coinbase utxos can only be sent to a zaddr.
+        errorString = ""
+        try:
+            self.nodes[0].sendtoaddress(self.nodes[2].getnewaddress(), 1.0)
+        except JSONRPCException,e:
+            errorString = e.error['message']
+        assert_equal("Coinbase funds can only be sent to a zaddr" in errorString, True)
+
+        # send node 0 taddr to node 0 zaddr
+        mytaddr = self.nodes[0].getnewaddress()
+        myzaddr = self.nodes[0].z_getnewaddress()
+        recipients = []
+        recipients.append({"address":myzaddr, "amount":20.0})
+        myopid = self.nodes[0].z_sendmany(mytaddr, recipients)
+        self.wait_for_operationd_success(myopid)
+        self.nodes[1].generate(1)
+        self.sync_all()
+
+        # check balances (the z_sendmany consumes 3 coinbase utxos)
+        resp = self.nodes[0].z_gettotalbalance()
+        assert_equal(Decimal(resp["transparent"]), Decimal('10.0'))
+        assert_equal(Decimal(resp["private"]), Decimal('29.9999'))
+        assert_equal(Decimal(resp["total"]), Decimal('39.9999'))
+
+        # convert note to transparent funds
+        recipients = []
+        recipients.append({"address":mytaddr, "amount":20.0})
+        myopid = self.nodes[0].z_sendmany(myzaddr, recipients)
+        self.wait_for_operationd_success(myopid)
+        self.nodes[1].generate(1)
+        self.sync_all()
+
+        # check balances
+        resp = self.nodes[0].z_gettotalbalance()
+        assert_equal(Decimal(resp["transparent"]), Decimal('30.0'))
+        assert_equal(Decimal(resp["private"]), Decimal('9.9998'))
+        assert_equal(Decimal(resp["total"]), Decimal('39.9998'))
+
+        # Send will fail because send amount is too big, even when including coinbase utxos
+        errorString = ""
+        try:
+            self.nodes[0].sendtoaddress(self.nodes[2].getnewaddress(), 99999.0)
+        except JSONRPCException,e:
+            errorString = e.error['message']
+        assert_equal("Insufficient funds" in errorString, True)
+
+        # Send will fail because of insufficient funds unless sender uses coinbase utxos
+        try:
+            self.nodes[0].sendtoaddress(self.nodes[2].getnewaddress(), 21.0)
+        except JSONRPCException,e:
+            errorString = e.error['message']
+        assert_equal("Insufficient funds, coinbase funds can only be spent after they have been sent to a zaddr" in errorString, True)
+
+        # Send will succeed because the balance of non-coinbase utxos is 20.0
+        errorString = ""
+        try:
+            self.nodes[0].sendtoaddress(self.nodes[2].getnewaddress(), 19.0)
+        except JSONRPCException:
+            assert(False)
+        assert_equal("" in errorString, True)
```

Comment:
You can/should also remove ``errorString = ""`` and the ``assert_equal``, since they have no effect.